### PR TITLE
Add GradSampler

### PR DESCRIPTION
**Pull request by KSGulin**
Add a GradSampler support to Yolov5

### CODE REVIEW

**Review by bfineran — COMMENTED** [review]
grad sampler lgtm, looks like you may need to rebase

**Review by bfineran — APPROVED** [review]

File: utils/sparse.py [hunk 1]
```diff
@@ -154,6 +172,29 @@ def reset_best(self, epoch):
 
         return (pruning_start <= epoch <= pruning_end) or epoch == qat_start
 
+    def _mfac_data_loader(self, train_loader, device, multi_scale, imgsz, gs):
+        def dataloader():
+            loader_type = type(train_loader)
+            mfac_dataloader = loader_type(
+                dataset=train_loader.dataset,
+                batch_size=train_loader.batch_size // 2,
```
Comment (dbogunowicz) [inline utils/sparse.py:180]: Why do we halve the batch size? Is it some property of MFAC?

Comment (KSGulin) [reply utils/sparse.py:180]: @dbogunowicz good question. The original MFAC paper recommended using half the batch size of training for MFAC calculations. We've decided not to include this as an exposed control for now

File: utils/sparse.py [hunk 2]
```diff
@@ -154,6 +172,29 @@ def reset_best(self, epoch):
...
+                sampler=deepcopy(train_loader.sampler),
+                num_workers=train_loader.num_workers,
+                collate_fn=train_loader.collate_fn,
+                pin_memory=train_loader.pin_memory,
+            )
+
+            for imgs, targets, *_ in mfac_dataloader:
```
Comment (rahul-tuli) [inline utils/sparse.py:187]: nice trick

**Review by rahul-tuli — APPROVED** [review]

**Comment by bfineran** [comment]
great work @KSGulin, landing

**Pull request merged**